### PR TITLE
Harden high-risk skill security boundaries

### LINKED ISSUE #198: [security] Fix insecure secret examples in page-monitoring
## Audit status

Snyk marks `page-monitoring` **Fail / High**:

- **W007 High:** insecure credential handling in examples.
- **W011 Medium:** arbitrary web/social content can influence alerts and archiving.

Details: https://skills.sh/jamditis/claude-skills-journalism/page-monitoring/security/snyk

The W007 report points to examples such as:

- `UptimeRobotClient('your-api-key')`
- inline Slack and Discord webhook strings
- an `email_config` dictionary containing SMTP username/password fields

These are placeholders, not leaked live credentials, but they teach readers to keep bearer secrets inline in source/config objects. Webhook URLs are credentials and should be handled as such.

The shared W011 remediation is tracked in #196.

## Acceptance criteria

- [ ] Replace inline secret placeholders with secure retrieval examples (environment variables, OS keyring/secret manager, or interactive configuration that does not persist plaintext).
- [ ] Validate that required secrets exist without printing their values.
- [ ] Document that webhook URLs, API keys, OAuth tokens, SMTP app passwords, and session files are secrets.
- [ ] Ensure logs, exception messages, diffs, alerts, and saved monitoring state redact credentials and authorization headers.
- [ ] Use least-privilege/service-specific credentials and document rotation/revocation.
- [ ] Keep secret-bearing files outside the repository with restrictive permissions and an explicit ignore rule.
- [ ] Add request timeouts/status checks and avoid including secrets in URL query strings where APIs support headers.
- [ ] Adopt #196 for monitored pages, RSS, and social content before that content can trigger alerts, commands, publication, or other writes.
- [ ] Add a repository check that rejects realistic secret values and inline credential assignments in public skill examples.
- [ ] Republish and request a fresh skills.sh audit.

### LINKED ISSUE #199: [security] Harden web-scraping trust boundaries and authenticated-session behavior
## Audit status

The current skills.sh page reports:

- Gen Agent Trust Hub: **Fail / High** — prompt injection, external downloads, and command execution.
- Socket: **Warn** — anti-bot evasion, stealth tooling, undocumented API discovery, and optional authenticated sessions.
- Snyk: **Warn / Medium** — W011 third-party content exposure.

Details:
- https://skills.sh/jamditis/claude-skills-journalism/web-scraping/security/agent-trust-hub
- https://skills.sh/jamditis/claude-skills-journalism/web-scraping/security/socket
- https://skills.sh/jamditis/claude-skills-journalism/web-scraping/security/snyk

## Review

One Gen finding is stale: it says `TrafilaturaСscraper` contains a Cyrillic homoglyph, while current source uses a Latin `C` in `TrafilaturaCscraper`. Renaming that class to the conventional `TrafilaturaScraper` would remove the ambiguity.

The remaining findings reflect real capabilities, but they are not evidence of malware. The skill intentionally launches browsers, downloads media, reads untrusted pages, discusses stealth, and can load an Instagram session file. The security problem is that these high-impact capabilities lack strong safety boundaries.

Repository-wide indirect prompt-injection work is tracked in #196.

## Acceptance criteria

- [ ] Rename `TrafilaturaCscraper` to `TrafilaturaScraper` and scan the file for non-ASCII identifier characters.
- [ ] Adopt the untrusted-content contract from #196: scraped content is data, never instructions; delimit it, cap it, preserve provenance, and gate follow-on actions.
- [ ] Validate requested URLs and redirects; reject unsafe schemes, loopback, link-local, and private-network destinations by default.
- [ ] Disable credentialed/authenticated scraping by default. Require explicit user approval, a named target, and the minimum necessary credential/session scope.
- [ ] Never print, embed, upload, or return cookies/session material.
- [ ] Separate ordinary extraction from bypass-oriented techniques and prefer official/research APIs when available.
- [ ] State that the skill must not defeat authentication, access controls, CAPTCHAs, or terms without documented authorization.
- [ ] Put subprocess/browser invocations behind explicit timeouts, output limits, and safe argv handling.
- [ ] Add tests/checks for Unicode confusables, unsafe URLs, redirects to private networks, and secret leakage.
- [ ] Republish and request a fresh skills.sh audit; record which residual warnings are expected for a scraping skill.

### LINKED ISSUE #203: web-scraping skill: four SSRF/redirect edge cases from the #201 review
Four findings from the review of #201 that are real but out of scope for that PR (it hardened the guidance text; these are follow-on gaps in the worked examples). Filing so they are on the record rather than dropped.

1. **Playwright follows redirects without re-validating the destination.** The browser example validates the initial URL, then hands it to Playwright, which will follow a redirect to a private-network address. The initial check does not cover where the navigation ends up.

2. **`requests` examples do not set `allow_redirects=False`.** Same class as (1) on the HTTP side: a validated public URL can redirect to `169.254.169.254` or an RFC1918 host, and the default follows it. The safe form validates each hop.

3. **The cascade does not distinguish a `401`/`403`/`429` from a transport failure.** The skill now says an access-denied response is a stop signal, but the example cascade falls through to the next tier on any exception, so an authorization block escalates to a heavier tool instead of stopping. The control is in the prose but not in the code.

4. **DNS-rebinding TOCTOU between validation and fetch.** Resolving a hostname to check it against the private-network ranges, then fetching by hostname, re-resolves and can land somewhere else. The mitigation is to pin the resolved address for the request, or to rely on network-level egress blocking rather than in-process validation (which the skill already recommends).

(1) and (2) are the same bug class and should be fixed together. (4) is why the skill's "run with private-network egress blocked" line is the load-bearing control and the in-process checks are defense in depth.

### LINKED ISSUE #205: Skill examples: webhook URLs leak through raise_for_status, plus two review nits
From the review of #201, filed rather than fixed in-PR (that PR's fix round is spent).

**1 and 2 — the real one, and it is the same class #201 was about.** `research-toolkit/skills/page-monitoring/SKILL.md:516` (Slack) and `:526` (Discord) call `response.raise_for_status()` on a webhook POST. `HTTPError`'s message carries the full request URL, and a webhook URL *is* the secret — there is no separate token. So an uncaught or logged failure prints the credential, in a skill that #201 just gave redaction guidance to. Catch `requests.RequestException` and re-raise without the URL or the response body.

Worth fixing both together: it is one pattern, twice.

**3 — `dev-toolkit/skills/web-scraping/SKILL.md:167`.** The Playwright context sets a generic browser User-Agent while the rest of the skill moves toward a stable descriptive one and says to identify the scraper. A generic UA there is mild misrepresentation and contradicts the surrounding text. Use the `ResearchScraper/1.0 ...` form.

**4 — `scripts/skill-security.test.mjs:69`.** The Unicode-confusable check blocks Cyrillic codepoints globally. It misses Greek/Coptic and other confusable ranges, and being global it also cannot distinguish an identifier from prose punctuation. Scanning `class`/`def`/assignment targets for non-ASCII would enforce the actual invariant instead of approximating it.

### PR DESCRIPTION
**Pull request by jamditis**
## What changed

- remove maintainer-specific OAuth token paths, hard-coded Drive destinations, and implicit upload behavior from `pdf-design`
- move page-monitoring secrets to fail-closed environment-backed configuration and redact URLs/errors in output
- add explicit untrusted-content, SSRF/private-network, access-control, and authenticated-session boundaries to `web-scraping`
- remove stealth/evasion guidance and the confusable scraper class typo
- validate every HTTP redirect, Playwright navigation, redirect, and subresource destination before fetching
- treat 401, 403, and 429 responses as stop signals instead of escalating the scraping cascade
- replace secret-bearing webhook exceptions with fixed redacted failures
- use a stable descriptive scraper user agent and scan declaration/assignment targets for non-ASCII identifiers
- add regression tests for the published security invariants

## Why

Skills.sh reported the affected skills at the highest current severities across Gen Agent Trust Hub and Snyk, with accompanying Socket warnings. The root causes were unsafe credential examples, maintainer-specific upload wiring, ambiguous trust boundaries, access-control-evasion guidance, redirect gaps, and secret-bearing exception text.

Closes #198
Closes #199
Closes #203
Closes #205
Addresses #200

Issue #200 spans both repositories and must remain open until this PR and jamditis/tools#81 have both landed.

## Impact

Published skill instructions now fail closed around credentials and authenticated sessions, require explicit upload destinations, treat downloaded/source content as untrusted, reject private-network scraping targets at every hop, and avoid leaking webhook credentials in failures.

## Validation

- `npm test` — 7/7 passing
- all 9 Python example blocks in `web-scraping/SKILL.md` compile
- `quick_validate.py` — all three edited skills valid
- `git diff --check` — clean

### CODE REVIEW

File: dev-toolkit/skills/web-scraping/SKILL.md [hunk 1]
```diff
@@ -1,11 +1,51 @@
 ---
 name: web-scraping
-description: Web scraping with anti-bot bypass, content extraction, undocumented APIs and poison pill detection. Use when extracting content from websites, handling paywalls, implementing scraping cascades or processing social media. Covers requests, trafilatura, Playwright with stealth mode, yt-dlp and instaloader patterns.
+description: Authorized web content extraction with trust-boundary controls, scraping cascades, poison-pill detection, browser rendering, observed API analysis, and social-media archiving. Use when extracting public content, diagnosing access failures, implementing respectful scrapers, or processing social-media sources with requests, trafilatura, Playwright, yt-dlp, or instaloader.
```
[the remark this answers is not in the document]
Comment (jamditis) [reply dev-toolkit/skills/web-scraping/SKILL.md:3]: Fixed in f582444. The page had drifted in six spots, not three: the three meta descriptions now carry the SKILL.md frontmatter verbatim so they cannot drift again, and the hero subhead ("anti-bot handling"), the "when to use" bullet ("Handling paywalls and anti-bot measures"), the tier-three fallback text, and the Playwright card were rewritten to describe diagnosing an access failure rather than defeating one. The hero and the "when to use" list were the worst of it -- they are the two things a reader sees before anything else, and they were still selling the behavior the skill now forbids. Swept the other two skills this PR hardened; only web-scraping had published docs that had drifted.

File: dev-toolkit/skills/web-scraping/SKILL.md [hunk 2]
```diff
@@ -119,10 +166,8 @@ class PlaywrightScraper(Scraper):
                 )
                 page = context.new_page()
 
-                # Apply stealth to avoid detection
-                stealth_sync(page)
-
                 page.goto(url, wait_until='networkidle', timeout=60000)
+                validate_public_url(page.url)
```
[the remark this answers is not in the document]
Comment (jamditis) [reply dev-toolkit/skills/web-scraping/SKILL.md]: Real, and out of scope for this PR -- #201 hardens the guidance text, while this is a gap in the worked examples. Filed as #203 with the detail so it is on the record rather than dropped. The redirect pair (Playwright and allow_redirects) is one bug class and should be fixed together; the DNS-rebinding TOCTOU is the reason the skill's "run with private-network egress blocked" line is the load-bearing control and the in-process URL checks are defense in depth.

File: dev-toolkit/skills/web-scraping/SKILL.md [hunk 3]
```diff
@@ -36,14 +75,22 @@ class Scraper(ABC):
...
-                return None
+            url = validate_public_url(url)
+            response = requests.get(
+                url,
+                headers={'User-Agent': 'ResearchScraper/1.0 (+https://example.org/contact)'},
+                timeout=30,
+                allow_redirects=False,
+            )
+            if response.is_redirect:
+                raise ValueError('Redirect target must be validated before fetching')
```
[the remark this answers is not in the document]
Comment (jamditis) [reply dev-toolkit/skills/web-scraping/SKILL.md]: Real, and out of scope for this PR -- #201 hardens the guidance text, while this is a gap in the worked examples. Filed as #203 with the detail so it is on the record rather than dropped. The redirect pair (Playwright and allow_redirects) is one bug class and should be fixed together; the DNS-rebinding TOCTOU is the reason the skill's "run with private-network egress blocked" line is the load-bearing control and the in-process URL checks are defense in depth.

File: dev-toolkit/skills/web-scraping/SKILL.md [hunk 4]
```diff
@@ -36,14 +75,22 @@ class Scraper(ABC):
...
+            response.raise_for_status()
```
[the remark this answers is not in the document]
Comment (jamditis) [reply dev-toolkit/skills/web-scraping/SKILL.md]: Real, and out of scope for this PR -- #201 hardens the guidance text, while this is a gap in the worked examples. Filed as #203 with the detail so it is on the record rather than dropped. The redirect pair (Playwright and allow_redirects) is one bug class and should be fixed together; the DNS-rebinding TOCTOU is the reason the skill's "run with private-network egress blocked" line is the load-bearing control and the in-process URL checks are defense in depth.

File: dev-toolkit/skills/web-scraping/SKILL.md [hunk 5]
```diff
@@ -36,14 +75,22 @@ class Scraper(ABC):
...
+class TrafilaturaScraper(Scraper):
     """Fast, lightweight extraction for standard articles."""
 
     def fetch(self, url: str) -> Optional[ScrapingResult]:
         try:
-            downloaded = trafilatura.fetch_url(url)
-            if not downloaded:
-                return None
+            url = validate_public_url(url)
+            response = requests.get(
```
[the remark this answers is not in the document]
Comment (jamditis) [reply dev-toolkit/skills/web-scraping/SKILL.md]: Real, and out of scope for this PR -- #201 hardens the guidance text, while this is a gap in the worked examples. Filed as #203 with the detail so it is on the record rather than dropped. The redirect pair (Playwright and allow_redirects) is one bug class and should be fixed together; the DNS-rebinding TOCTOU is the reason the skill's "run with private-network egress blocked" line is the load-bearing control and the in-process URL checks are defense in depth.

File: research-toolkit/skills/page-monitoring/SKILL.md [hunk 6]
```diff
@@ -470,14 +512,18 @@ class AlertManager:
         if channel:
             payload['channel'] = channel
 
-        requests.post(self.slack_webhook, json=payload)
+        response = requests.post(self.slack_webhook, json=payload, timeout=15)
+        response.raise_for_status()
```
[the remark this answers is not in the document]
Comment (jamditis) [reply research-toolkit/skills/page-monitoring/SKILL.md]: Filed as #205 rather than fixed here — this PR's fix round is spent. The webhook pair is the one that matters and they are one pattern twice: a webhook URL is itself the secret, so raise_for_status leaking it in an HTTPError message is exactly the class this PR added redaction guidance for.

File: research-toolkit/skills/page-monitoring/SKILL.md [hunk 7]
```diff
@@ -470,14 +512,18 @@ class AlertManager:
...
     def send_discord(self, message: str):
         """Send Discord notification."""
         if not self.discord_webhook:
             return
 
-        requests.post(self.discord_webhook, json={'content': message})
+        response = requests.post(
+            self.discord_webhook, json={'content': message}, timeout=15
+        )
+        response.raise_for_status()
```
[the remark this answers is not in the document]
Comment (jamditis) [reply research-toolkit/skills/page-monitoring/SKILL.md]: Filed as #205 rather than fixed here — this PR's fix round is spent. The webhook pair is the one that matters and they are one pattern twice: a webhook URL is itself the secret, so raise_for_status leaking it in an HTTPError message is exactly the class this PR added redaction guidance for.

File: dev-toolkit/skills/web-scraping/SKILL.md [hunk 8]
```diff
@@ -119,10 +166,8 @@ class PlaywrightScraper(Scraper):
                 )
                 page = context.new_page()
```
[the remark this answers is not in the document]
Comment (jamditis) [reply dev-toolkit/skills/web-scraping/SKILL.md:193]: Filed as #205 rather than fixed here — this PR's fix round is spent. The webhook pair is the one that matters and they are one pattern twice: a webhook URL is itself the secret, so raise_for_status leaking it in an HTTPError message is exactly the class this PR added redaction guidance for.

File: scripts/skill-security.test.mjs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+  assert.match(skill, /never (log|print).*secret/i);
+  assert.match(skill, /type\(error\)\.__name__/);
+});
+
+test("web-scraping defines explicit content, URL, and session trust boundaries", async () => {
+  const skill = await readFile(webScrapingUrl, "utf8");
+
+  assert.doesNotMatch(skill, /TrafilaturaCscraper/);
+  assert.doesNotMatch(skill, /[\u0400-\u04ff]/);
+  assert.match(skill, /class TrafilaturaScraper/);
```
[the remark this answers is not in the document]
Comment (jamditis) [reply scripts/skill-security.test.mjs:88]: Filed as #205 rather than fixed here — this PR's fix round is spent. The webhook pair is the one that matters and they are one pattern twice: a webhook URL is itself the secret, so raise_for_status leaking it in an HTTPError message is exactly the class this PR added redaction guidance for.
